library/gcm.c

```diff
@@ -332,28 +332,53 @@ int mbedtls_gcm_starts( mbedtls_gcm_context *ctx,
     return( 0 );
 }
 
-
+/**
+ * mbedtls_gcm_context::buf contains different data type, depending
+ * on the values of mbedtls_gcm_context::::add_len and
+ * mbedtls_gcm_context::len:
+ *     * When add_len % 16 == 0 and len == 0: initial state.
```

Comment:
Looking at this again I think it would be best explained with four cases:
* If `len == 0 && add_len % 16 == 0`: the first `add_len % 16 == 0` bytes have a partial block of AD that has been xored in but not yet multiplied in.
* If `len == 0 && add_len % 16 == 0`: the authentication tag is correct if the data ends now.
* If `len % 16 != 0`: the first `len % 16 == 0` bytes have a partial block of ciphertext that has been xored in but not yet multiplied in.
* If `len > 0 && len % 16 == 0`: the authentication tag is correct if the data ends now.